tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -1508,6 +1508,69 @@ SPIRVEmitter::processStructuredBufferLoad(const CXXMemberCallExpr *expr) {
   return theBuilder.createAccessChain(ptrType, doExpr(buffer), {zero, index});
 }
 
+uint32_t
+SPIRVEmitter::processACSBufferAppendConsume(const CXXMemberCallExpr *expr) {
+  const bool isAppend = expr->getNumArgs() == 1;
+
+  const uint32_t u32Type = theBuilder.getUint32Type();
+  const uint32_t one = theBuilder.getConstantUint32(1);  // As scope: Device
+  const uint32_t zero = theBuilder.getConstantUint32(0); // As memory sema: None
+
+  const auto *object = expr->getImplicitObjectArgument();
+  const auto *buffer = cast<DeclRefExpr>(object)->getDecl();
+
+  // Calculate the index we should use for appending the value
+  const uint32_t counterVar = declIdMapper.getCounterId(cast<VarDecl>(buffer));
+  const uint32_t counterPtrType = theBuilder.getPointerType(
+      theBuilder.getInt32Type(), spv::StorageClass::Uniform);
+  const uint32_t counterPtr =
+      theBuilder.createAccessChain(counterPtrType, counterVar, {zero});
+  uint32_t index = 0;
+  if (isAppend) {
+    // For append, we add one to the counter.
+    index = theBuilder.createAtomicIAdd(u32Type, counterPtr, one, zero, one);
+  } else {
+    // For consume, we substract one from the counter. Note that OpAtomicIAdd
+    // returns the value before the addition; so we need to do substraction
+    // again with OpAtomicIAdd's return value.
+    const uint32_t negOne = theBuilder.getConstantInt32(-1);
+    const auto prevIndex =
+        theBuilder.createAtomicIAdd(u32Type, counterPtr, one, zero, negOne);
```

Comment:
is there a reason you chose not to use `OpAtomicISub` with `one`?

Reply:
Good catch! Somehow it didn't come to me that I can use subtraction. Ugh.